### ISSUE TITLE
az container commands throwing error

### ISSUE THREAD

**Issue by myusrn** [Azure/azure-cli#8123]
**Describe the bug**
When i execute "az container --help" or any other container command, e.g. "az container create -g <resource group name> -n <container name> --image <public registry image name[:tag]> --ip-address public --ports <public access port>" using my localhost azure-cli ( 2.0.50 ) install it throws a bunch of error output the first line of it being " No module named 'azure.mgmt.containerinstance' ".  If i execute "pip list | findstr /r /i /c:"^azure-mgmt-containerinstance" there is a "azure-mgmt-containerinstance          1.0.0" found.  

**To Reproduce**
1. pip install --user azure-cli
2. az container --help
3. output is 
```
No module named 'azure.mgmt.containerinstance'
Traceback (most recent call last):
  File "C:\Users\ob1\AppData\Roaming\Python\Python36\site-packages\knack\cli.py", line 197, in invoke
    cmd_result = self.invocation.execute(args)
  File "c:\users\ob1\appdata\roaming\python\python36\site-packages\azure\cli\core\commands\__init__.py", line 286, in execute
    parsed_args = self.parser.parse_args(args)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\argparse.py", line 1730, in parse_args
    args, argv = self.parse_known_args(args, namespace)
</snip>
```

**Expected behavior**
expected help details for container command set like i get if i execute "az container . . ." commands in cloud shell

**Environment summary**
Windows [Version 10.0.17763.195]
Python 3.6.5 :: Anaconda, Inc.
azure-cli (2.0.50)  [ installed using "pip install --user azure-cli" ]

**Additional context**
While i can use the azure portal's cloud shell to execute these commands i would like to, and assumed i could, be able to run these using localhost azure-cli install.

**Comment by myusrn** [comment]
@williexu thanks for the followup.  Can you clarify how I should go about getting the current v2.0.54 install update?  I ask because when I run "`pip install --user azure-cli`" and afterwards do an "`pip list`" and "`az.bat --version`" the both are still showing v2.0.50.

**Comment by williexu** [comment]
@myusrn something went very wrong with your install, most of the dependencies are outdated (for example, `azure-cli-core` should also be at 2.0.54. 
Can you try using `pip install --upgrade --upgrade-strategy eager`?

I would also actually recommend installing through the right system installer if possible: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest and if you want to remain using pip, to do so in a new python virtual environment.

**Comment by myusrn** [comment]
@williexu I executed `pip install --upgrade --upgrade-strategy eager --user azure-cli` and this produced `pip list | findstr /r /i /c:"^azure-cli"` and `az --version` output that matches what I get if I use `pip uninstall azure-cli` from administrator command prompt window followed by `pip install --user azure-cli`, attached zip of these outputs.  

Even with those updated, or clean install, dependencies in place executing `az container --help` still produced the error messages output beginning with `No module named 'azure.mgmt.containerinstance'` 

I did another `pip uninstall azure-cli` from administrator command prompt, confirming afterwards that `dir %appdata%\python\python36\scripts\az*` had no file results and `for /D %i in ( %appdata%\python\python36\site-packages\azure* ) do ( rd /s /q %i ) & dir %appdata%\python\python36\site-packages\azure*` had no folder results, followed by applying the https://aka.ms/installazurecliwindows | azure-cli-2.0.54.msi 20dec18 windows installer option.  

With that change in place `az container --help` is now succeeding.

**Comment by williexu** [comment]
@myusrn glad you were able to install properly using the msi. We also have installers for homebrew and various linux distros.

If you are installing through pip, please try creating a python virtual environment and doing a fresh pip install azure-cli. I just tried on windows using pip v18.1 with no problems.

**Comment by myusrn** [comment]
@williexu thanks for additional details and results of your tests using pip against are a python virtual environment.  

Are you thinking that the visual studio 2017 [ / 2019 ] workload installers that create the "%programfiles(x86)%\Microsoft Visual Studio\Shared\Anaconda3_64" python 3.6.5 & pip 18.1 install I was trying to use is for some reason not handling the "pip install azure-cli" installation correctly?

**Comment by tjprescott** [comment]
TL/DR: use virtual environments any time you `pip install azure-cli`.

The ways in which pip can fail spectacularly when trying to update the CLI is why we simply do not recommend direct pip install of the CLI unless it is in a virtual environment. The virtual environment effectively makes uninstall a snap since you can just throw away the entire environment and start anew. To make matters worse, the ways in which you can screw up you installation depend on the version of pip you have installed and make it nearly impossible for us to assist with.

In general, you need to `pip uninstall` all of the azure-cli-* files. Simply running `pip uninstall azure-cli` doesn't uninstall all the other files in the way that `pip install azure-cli` installs them. Sometimes you'll run into issues where, after uninstalling something, you will run `pip list` and still see that package listed, albeit at a different location that doesn't seem to respond to `pip uninstall`.  In that case, you need to go to that directory and remove the `dist-info` or `egg-info` folder for pip to no longer detect it. 

As @williexu points out, the packaged installers will ensure you have all the correct dependencies and their uninstall/update mechanisms work much more reliably, which is why we recommend them.

Since the question has been answered, I will close this issue.

**Comment by myusrn** [comment]
@tjprescott tahnks for the additional details.  Yes I am finding that backing out of a `pip install --user azure-cli` using `pip uninstall azure-cli` still leaves behind all the `%appdata%\Python\Python36\site-packages\azure*` folders and in the case of a `pip install azure-cli` install it still leaves behind all the `%programfiles(x86)%\Microsoft Visual Studio\Shared\Anaconda3_64\pkgs\azure*` folders.  I used `for /D %i in ( azure*.dist-info azure*.egg-info ) do ( rd /s /q %i )` from those user and system install packages folders to clean them up.  Afterwards I find that `pip list | findstr /r /i /c:"azure"` still returns some result with `azure-mgmt-*` and `azure-nspkg-*` naming so not sure if those are related or different python packages installed and used by vstudio and/or vscode.

**Comment by tjprescott** [comment]
The `azure-mgmt-*` and `azure-nspkg-*` files are the SDK. You can uninstall those as well, though you might have other things that need them. I assume that's why pip doesn't uninstall things the same way that it installed them. It's easier to install dependencies than for pip to figure out "this was a dependency and now nothing else uses it so uninstall it too". However, uninstalling the packaged versions will take care of this since each packaged version maintains its dependencies in a separate space.
